### ISSUE TITLE
"Edit Event Series" functionality cannot save changes to Event dates/times; invisible 500 error, Drupal provides no user feedback

### ISSUE THREAD

**Issue by stefaniefgray** [department-of-veterans-affairs/va.gov-cms#12529]
## Describe the defect
When editing a recurring event, the "Edit Event Series" functionality cannot save any changes made to individual event dates/times.

### User story:
As an editor editing an Event that happens every Tuesday, I would like to change the February 14th event to February 15th.

If I click on "Edit Event Series," I can edit the February 14th event and change it to February 15th. **However, if I click "Save," nothing happens, and the site is unresponsive.**

I can click "Cancel" or X out of the pop-up window, but I cannot save my changes to the event date/time, or any event within the series listing.

## To Reproduce

1. Go to [an existing event](https://staging.cms.va.gov/admin/content?title=&type=event&moderation_state=All&owner=All).
2. Click “Edit” and make sure the event is marked as recurring, with a specific frequency provided (i.e. Every 2 weeks on Wednesdays until 12/31/2024).
3. Click "Save" at the bottom of the page.
4. Edit the event once more, go to the "Recurring Events" settings, and click the “Edit Event Series” button.
5. You will see a pop-up window with all of the upcoming event dates listed. Click "Edit" next to one of the event occurrences, and change the date and/or time of the Event. When complete, click "Save."
6. After clicking "Save," Drupal gets stuck. You can click “Cancel” or X out of the pop-up, but it won’t save your changes to the event series.

**Note:** If you do cancel or X out of the pop-up window and click "Edit Event Series" once more, Drupal also gets stuck and will not open the pop-up window again.

## AC / Expected behavior
Editors on https://prod.cms.va.gov/ can change the date and/or time of a specific event occurrence within a recurring series without Drupal getting stuck.

#### Listing of recurring event dates:
![Screen Shot 2023-02-08 at 2 50 01 PM](https://user-images.githubusercontent.com/9059428/217666487-81baf10a-f159-4ea1-aa9a-49cdd7dd0e35.png)

#### Screenshot of recurring event settings and "Edit Event Series" button:
![Screen Shot 2023-02-08 at 2 50 56 PM](https://user-images.githubusercontent.com/9059428/217666590-8f630346-f33d-4560-a5a7-5903f30b432c.png)

#### "Manage Event series" pop-up window with "Edit" button:
![Screen Shot 2023-02-08 at 2 51 08 PM](https://user-images.githubusercontent.com/9059428/217666706-549c1f49-7ae4-43d4-b800-ecc0e3bef290.png)

#### "Manage Event series" pop-up window when changing Feb. 14th to Feb 15th -- this is the location of the "Save" button that gets stuck and makes the site unresponsive:
![Screen Shot 2023-02-08 at 2 51 43 PM](https://user-images.githubusercontent.com/9059428/217666821-c135c441-d650-48dc-afb8-2a4cb9c9935d.png)

#### AJAX error
On save when testing on a test CMS instance, Console shows the following errors: 
```
POST https://pr12421-1tnucvlbg0mccusqcjkatzs4qbc4qa7j.ci.cms.va.gov/admin/content/smart_date_recur/246/instance/reschedule/3/ajax?_wrapper_format=drupal_ajax&ajax_form=1&_wrapper_format=drupal_ajax 500
  send @ jquery.min.js?v=3.6.3:2
  ajax @ jquery.min.js?v=3.6.3:2
  q.fn.ajaxSubmit @ jquery.form.js:341
  Drupal.Ajax.eventResponse @ ajax.js?v=9.5.2:293
  (anonymous) @ ajax.js?v=9.5.2:247
  dispatch @ jquery.min.js?v=3.6.3:2
  y.handle @ jquery.min.js?v=3.6.3:2
```

> Uncaught Drupal.AjaxError
> '\nAn AJAX HTTP error occurred.\nHTTP Result Code: 500\nDebugging information follows.\nPath: /admin/content/smart_date_recur/246/instance/reschedule/3/ajax?_wrapper_format=drupal_ajax&ajax_form=1\nStatusText: error\nResponseText: The website encountered an unexpected error. Please try again later.Drupal\\Core\\Entity\\EntityStorageException: Route &quot;entity.smart_date_override.canonical&quot; does not exist. in Drupal\\Core\\Entity\\Sql\\SqlContentEntityStorage-&gt;save() (line 815 of core/lib/Drupal/Core/Entity/Sql/SqlContentEntityStorage.php).'

 ## Additional context
Reported in https://va-gov.atlassian.net/browse/VAHELP-5367

Slack thread: https://dsva.slack.com/archives/C52CL1PKQ/p1675886683643809

## Labels
(You can delete this section once it's complete)
- [x] Issue type (red) (defaults to "Defect")
- [ ] CMS subsystem (green)
- [ ] CMS practice area (blue)
- [x] CMS workstream (orange) (not needed for bug tickets)
- [ ] CMS-supported product (black)

### Team
Please check the team(s) that will do this work.

- [ ] `CMS Team`
- [x] `Public Websites`
- [ ] `Facilities`
- [x] `User support`

**Comment by jilladams** [comment]
Added AJAX error to the ticket body: 
![Uploading Screen Shot 2023-02-08 at 2.55.19 PM.png…]()

**Comment by jilladams** [comment]
Assigned to @dsasser for pre-refinement, in case there's anything we can learn before planning to help get a sense of sizing. (It's a bug, so maybe not.)

**Comment by ndouglas** [comment]
Likely tied to this: #12444 .

**Comment by ndouglas** [comment]
Note that `smart_date` isn't up-to-date -- I actually updated it from 3.5.1 -> 3.6.0 some time back and it broke things because they changed something in their schema.  I opened an issue for dealing with that [here](https://github.com/department-of-veterans-affairs/va.gov-cms/issues/10490).

**Comment by dsasser** [comment]
I was able to reproduce this locally. The full error is:

```
An AJAX HTTP error occurred.
HTTP Result Code: 500
Debugging information follows.
Path: /admin/content/smart_date_recur/154/instance/reschedule/10/ajax?_wrapper_format=drupal_ajax&ajax_form=1
StatusText: error
ResponseText: The website encountered an unexpected error. Please try again later.Drupal\Core\Entity\EntityStorageException: Route &quot;entity.smart_date_override.canonical&quot; does not exist. in Drupal\Core\Entity\Sql\SqlContentEntityStorage-&gt;save() (line 815 of core/lib/Drupal/Core/Entity/Sql/SqlContentEntityStorage.php)."
```

**Comment by dsasser** [comment]
I found the root cause of this issue: the `SmartDateOverride` entity type has a `canonical` link template, but is missing the corresponding route.

This problem gets exposed when code attempts to use the route, such as when creating a route url to the canonical link. Since the link template exists, but not the route, the error is not generally revealed under "normal" circumstances. In our case, however, we are using a module, `next`, that does attempt to generate the Url of entities during an entity_insert hook. The `next` module's entity_insert hook executes EntityActionEvent::createFromEntity() which does the following:

```
  public static function createFromEntity(EntityInterface $entity, string $action): self {
    /** @var \Drupal\next\NextEntityTypeManagerInterface $next_entity_type_manager */
    $next_entity_type_manager = \Drupal::service('next.entity_type.manager');

    $sites = $next_entity_type_manager->getSitesForEntity($entity);
    $url = $entity->hasLinkTemplate('canonical') ? $entity->toUrl()->toString() : NULL;

    return new static($entity, $action, $sites, $url);
  }
```

Here, the `next` module is checking if the entity has a `canonical` link template, and generating a url from the entity if so. However, since the route doesn't exist, there is an error thrown when `$entity->toUrl()->toString() ` is called.

The latest smart_date module version does not solve this problem.

**Comment by dsasser** [comment]
There is an additional 500 error that can occur when nodes are not "attached" to the smart date recurrence rule entity. This was discovered during[ PR review of this issue](https://github.com/department-of-veterans-affairs/va.gov-cms/pull/12567#issuecomment-1426258544).

**Comment by jilladams** [comment]
@dsasser I know there's future work / bigger pieces in #12568, but for now: is @stefaniefgray ok to verify and close loop with customer / we can close this ticket?

**Comment by dsasser** [comment]
> @dsasser I know there's future work / bigger pieces in #12568, but for now: is @stefaniefgray ok to verify and close loop with customer / we can close this ticket?

Yes the issues reported by Stefanie are now resolved on prod and she can close the loop with customers.

**Comment by dsasser** [comment]
Bug tested and passed on production.

**Comment by stefaniefgray** [comment]
@dsasser You rock Daniel, thank you!!! And thank you @jilladams for all of your work on this ticket as well!
